File: 74-search-a-2d-matrix/search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        # find candidate row
        # perform binary search on matrix

        def isCandidateRow(array, val):
            i = 0
            j = len(array) - 1
            return array[i] <= val <= array[j]

        def binary_search(array, val):
            i = 0
            j = len(array) - 1

            while (i <= j):
                mid = (i + j) // 2
                if (array[mid] == val):
                    return True
                if (val < array[mid]):
                    j = mid - 1
                else:
                    i = mid + 1

            return False
        
        i = 0
        j = len(matrix) - 1

        while (i <= j):
            mid = (i + j) // 2

            if (isCandidateRow(matrix[mid], target)):
                return binary_search(matrix[mid], target)
            if (target < matrix[mid][0]):
                j = mid - 1
            else:
                i = mid + 1

        
        return False

        

File: 74-search-a-2d-matrix/search-a-2d-matrix.js
var searchMatrix = function(matrix, target) {
    function isInRange(val, array) {
        let i = 0
        let j = array.length - 1
        return array[i] <= val && val <= array[j]
    }

    function binarySearch(val, array) {
        let i = 0
        let j = array.length - 1

        while (i <= j) {
            let mid = Math.floor((i + j) / 2)
            if (array[mid] == val) return true
            else if (val < array[mid]) j = mid - 1
            else i = mid + 1
        }

        return false
    }

    let i = 0
    let j = matrix.length - 1

    while (i <= j) {
        let mid = Math.floor((i + j) / 2)
        if (isInRange(target, matrix[mid])) return binarySearch(target, matrix[mid])
        else if (target < matrix[mid][0]) j = mid - 1
        else i = mid + 1
    }

    return false
};